src/OpenTelemetry.Api/Trace/SpanAttribute.cs

```diff
@@ -0,0 +1,121 @@
+﻿// <copyright file="SpanAttribute.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace OpenTelemetry.Trace
+{
+    /// <summary>
+    /// A class that represents the span attributes.
+    /// </summary>
+    public struct SpanAttribute
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SpanAttribute"/> struct.
+        /// </summary>
+        /// <param name="initialAttributes">Initial attributes for the <see cref="SpanAttribute"/>.</param>
+        public SpanAttribute(IEnumerable<KeyValuePair<string, object>> initialAttributes)
+        {
+            this.Attributes = initialAttributes == null ? new ActivityTagsCollection() : new ActivityTagsCollection(initialAttributes);
+        }
+
+        /// <summary>
+        /// Gets Attributes.
+        /// </summary>
+        public ActivityTagsCollection Attributes { get; }
```

Comment:
Do we need this?

Reply:
please make it internal

Reply:
> Do we need this?

I think so. Did you imagine another way to achieve this?

Reply:
Hmm, if this class is meant to be a light wrapper around an `ActivityTagsCollection` then it does make sense for this to be internal (maybe private?). Does it make sense for this class to implement `IDictionary<string, object>` and then just be a pass-through to the underlying `ActivityTagsCollection`. See: https://github.com/dotnet/runtime/blob/b89566689612a36e0c44e55597677dd230c4f98a/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs#L23

Reply:
before this change we were using the IEnumerable<KeyValue<T, U>> as parameter. The idea to change to ActivityTagsCollection is to prevent the conversion between one/another. That was my understanding.

I'm not sure.

@cijothomas , what do you think?

Reply:
Just thinking this out a bit more... how about something like:

```csharp
public class SpanAttributes : IDictionary<string, object>
{
    private ActivityTagsCollection activityTagsCollection = new ActivityTagsCollection();

    public SpanAttributes()
    {
        activityTagsCollection = new ActivityTagsCollection();
    }

    public SpanAttributes(IEnumerable<KeyValuePair<string, object>> initialAttributes)
    {
         activityTagsCollection = new ActivityTagsCollection(initialAttributes);
    }

    // implement complete shim interface to underlying ActivityTagsCollection here
}
```

Reply:
since i created as struct, we can't have a struct with no parameters in constructor. But, don't know if we should go as struct or as class.

Reply:
Not sure I see the need for this to be a struct. Don't quote me, but I'm not sure I've ever encountered a collection data structure be represented as a struct in C#. Structs normally represent immutable data. Do you have more background information regarding your choice of a struct?

Reply:
I used struct because other parts of the Span were as well. So it's ok to change :)

Reply:
usage of struct here is to avoid the heap allocation.

Reply:
should i change to class and do as @alanwest suggested or should i maintain this?

Reply:
A struct is perfectly fine. You'll just have to deal with the initialization of the backing ActivityTags collection differently. @cijothomas has a suggestion below.

My initial gut was that the heap allocation overhead would be minimal, but with nothing to back that up - and knowing intuitively that this would be a hot path for a consumer of the shim API, a struct makes sense. Probably makes sense to get some benchmarks in place for the shim API at some point.

Reply:
It looks like you've changed it to a class, but in case we still want to entertain a struct, does this help?

```csharp
private ActivityTagsCollection ActivityTagsCollection => this.activityTagsCollection ??= new ActivityTagsCollection();
```

Then eliminate the constructor that takes `initialAttributes`. We could introduce an `AddRange` that would do all the adds in conformance with the spec.

```csharp
public void AddRange(IEnumerable<KeyValuePair<string, object>> initialAttributes)
```

🤷 Does this feel right to folks?

Reply:
i will change tomorrow! thank you for the suggestion

Reply:
just updated.